File: src/extras-kcoreaddons/libkjsonutils.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kjsonutils.html
pub const kjsonutils = struct {
    /// [Qt documentation](https://api.kde.org/kjsonutils.html#readTranslatedValue)
    ///
    /// ``` param1: QtC.QJsonObject, param2: []const u8, param3: QtC.QJsonValue ```
    pub fn ReadTranslatedValue(param1: ?*anyopaque, param2: []const u8, param3: ?*anyopaque) QtC.QJsonValue {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        return qtc.KJsonUtils_ReadTranslatedValue(@ptrCast(param1), param2_str, @ptrCast(param3));
    }

    /// [Qt documentation](https://api.kde.org/kjsonutils.html#readTranslatedString)
    ///
    /// ``` param1: QtC.QJsonObject, param2: []const u8, param3: []const u8, allocator: std.mem.Allocator ```
    pub fn ReadTranslatedString(param1: ?*anyopaque, param2: []const u8, param3: []const u8, allocator: std.mem.Allocator) []const u8 {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        const param3_str = qtc.libqt_string{
            .len = param3.len,
            .data = param3.ptr,
        };
        const _str = qtc.KJsonUtils_ReadTranslatedString(@ptrCast(param1), param2_str, param3_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kjsonutils.ReadTranslatedString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }
};

File: src_original/extras-kcoreaddons/libkjsonutils.cpp
#include <KJsonUtils>
#include <QJsonObject>
#include <QJsonValue>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <kjsonutils.h>
#include "libkjsonutils.h"
#include "libkjsonutils.hxx"

QJsonValue* KJsonUtils_ReadTranslatedValue(const QJsonObject* param1, const libqt_string param2, const QJsonValue* param3) {
    QString param2_QString = QString::fromUtf8(param2.data, param2.len);
    return new QJsonValue(KJsonUtils::readTranslatedValue(*param1, param2_QString, *param3));
}

libqt_string KJsonUtils_ReadTranslatedString(const QJsonObject* param1, const libqt_string param2, const libqt_string param3) {
    QString param2_QString = QString::fromUtf8(param2.data, param2.len);
    QString param3_QString = QString::fromUtf8(param3.data, param3.len);
    QString _ret = KJsonUtils::readTranslatedString(*param1, param2_QString, param3_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}
